File: user-service/src/main/java/rs/raf/user_service/controller/AuthController.java
package rs.raf.user_service.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import rs.raf.user_service.domain.dto.*;
import rs.raf.user_service.service.AuthService;

@Tag(name = "Authentication Controller", description = "API for authenticating users")
@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @Operation(summary = "Client Login", description = "Endpoint for logging client and generating JWT token")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Succesfully generated JWT token"),
            @ApiResponse(responseCode = "401", description = "Bad credentials")
    })
    @PostMapping("/login/client")
    public ResponseEntity<?> clientLogin(@RequestBody LoginRequestDto request) {
        String token = authService.authenticateClient(request.getEmail(), request.getPassword());
        if (token == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Bad credentials");
        }
        LoginResponseDto response = new LoginResponseDto();
        response.setToken(token);
        return ResponseEntity.ok(response);
    }

    @Operation(summary = "Employee Login", description = "Endpoint for logging employee and generating JWT token")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Succesfully generated JWT token"),
            @ApiResponse(responseCode = "401", description = "Bad credentials")
    })
    @PostMapping("/login/employee")
    public ResponseEntity<?> employeeLogin(@RequestBody LoginRequestDto request) {
        String token = authService.authenticateEmployee(request.getEmail(), request.getPassword());
        if (token == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Bad credentials");
        }
        LoginResponseDto response = new LoginResponseDto();
        response.setToken(token);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/request-password-reset")
    @Operation(summary = "Request password reset", description = "Requests password reset with email adress.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Request for password reset successfully sent."),
            @ApiResponse(responseCode = "400", description = "Invalid email.")
    })
    public ResponseEntity<Void> requestPasswordReset(@RequestBody RequestPasswordResetDto requestPasswordResetDTO) {
        try {
            this.authService.requestPasswordReset(requestPasswordResetDTO.getEmail());
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PostMapping("/check-token")
    @Operation(summary = "Checks if a token is still valid", description = "Checks if a token is still valid.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Valid"),
            @ApiResponse(responseCode = "404", description = "Invalid")
    })
    public ResponseEntity<Void> checkToken(@RequestBody CheckTokenDto checkTokenDTO) {
        try {
            authService.checkToken(checkTokenDTO.getToken());
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(404).build();
        }
    }

    @PostMapping("/request-card")
    @Operation(summary = "Request card", description = "Requests card.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Request for password reset successfully sent."),
            @ApiResponse(responseCode = "400", description = "Invalid email.")
    })
    public ResponseEntity<Void> requestCard(@RequestBody RequestCardDto requestCardDto) {
        try {
            this.authService.requestCard(requestCardDto.getEmail());
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PostMapping("/set-password")
    @Operation(summary = "Sets password", description = "Sets new password for both clients and employees")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Password set successfully."),
            @ApiResponse(responseCode = "400", description = "Invalid data.")
    })
    public ResponseEntity<Void> activateUser(@RequestBody ActivationRequestDto activationRequestDto) {
        try {
            authService.setPassword(activationRequestDto.getToken(), activationRequestDto.getPassword());
            //userService.activateUser(activationRequestDto.getToken(), activationRequestDto.getPassword());
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/controller/AuthController.kt
package pack.userservicekotlin.controller

import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import pack.userservicekotlin.arrow.AuthServiceError
import pack.userservicekotlin.domain.dto.activity_code.ActivationRequestDto
import pack.userservicekotlin.domain.dto.activity_code.RequestPasswordResetDto
import pack.userservicekotlin.domain.dto.external.CheckTokenDto
import pack.userservicekotlin.domain.dto.login.LoginRequestDto
import pack.userservicekotlin.domain.dto.login.LoginResponseDto
import pack.userservicekotlin.service.AuthService
import pack.userservicekotlin.swagger.AuthApiDoc

@Tag(name = "Authentication Controller", description = "API for authenticating users")
@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val authService: AuthService,
) : AuthApiDoc {
    @PostMapping("/login/client")
    override fun clientLogin(
        @RequestBody request: LoginRequestDto,
    ): ResponseEntity<Any> =
        authService.authenticateClient(request.email, request.password).fold(
            ifLeft = {
                when (it) {
                    AuthServiceError.InvalidCredentials -> ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Bad credentials")
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { token ->
                val response = LoginResponseDto().apply { this.token = token }
                ResponseEntity.ok(response)
            },
        )

    @PostMapping("/login/employee")
    override fun employeeLogin(
        @RequestBody request: LoginRequestDto,
    ): ResponseEntity<Any> =
        authService.authenticateEmployee(request.email, request.password).fold(
            ifLeft = {
                when (it) {
                    AuthServiceError.InvalidCredentials -> ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Bad credentials")
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { token ->
                val response = LoginResponseDto().apply { this.token = token }
                ResponseEntity.ok(response)
            },
        )

    @PostMapping("/api/auth/request-password-reset")
    override fun requestPasswordReset(
        @RequestBody requestPasswordResetDTO: RequestPasswordResetDto,
    ): ResponseEntity<Void> =
        authService.requestPasswordReset(requestPasswordResetDTO.email).fold(
            ifLeft = {
                when (it) {
                    AuthServiceError.UserNotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).build()
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PostMapping("/api/auth/check-token")
    override fun checkToken(
        @RequestBody checkTokenDTO: CheckTokenDto,
    ): ResponseEntity<Void> =
        authService.checkToken(checkTokenDTO.token).fold(
            ifLeft = {
                when (it) {
                    AuthServiceError.InvalidToken, AuthServiceError.ExpiredToken -> ResponseEntity.status(HttpStatus.NOT_FOUND).build()
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PostMapping("/api/auth/set-password")
    override fun activateUser(
        @RequestBody activationRequestDto: ActivationRequestDto,
    ): ResponseEntity<Void> =
        authService.setPassword(activationRequestDto.token, activationRequestDto.password).fold(
            ifLeft = {
                when (it) {
                    AuthServiceError.InvalidToken, AuthServiceError.ExpiredToken, AuthServiceError.UserNotFound ->
                        ResponseEntity.status(HttpStatus.NOT_FOUND).build()
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok().build() },
        )
}
